Add explanatory info to intervention examples (#758)
cylc/cylc-doc: src/user-guide/interventions/index.rst

Sometimes things don't go to plan!

> So Cylc allows you to take manual control of your workflow whilst it's running,

Reviewer: ```suggestion
Cylc allows you to take manual control of your workflow whilst it's running,
```

> So Cylc allows you to take manual control of your workflow whilst it's running,
> to do things like edit a task's configuration, re-run a section

Reviewer: If we can get rid of "do things like" and make the list happily comprehensive I'd be happier.

```suggestion
Cylc allows you to take manual control of your workflow whilst it's running,
to do things like edit a task's configuration or re-run a section
```

Author: I was trying to keep this PR strictly to "explanatory info for the examples", rather than a more general revision (note there's a "whilst" in there too 😁 ) ... but I tend to agree with you on that point.

         $ cylc trigger <workflow>//<task>

   .. tab-item:: Explanation

>       :sync: explanation

Reviewer 2: IMO this should not be a tab, which is meant to be for your preferred choice of interacting with Cylc. Instead it should be a "spoiler" admonition or ordinary admonition

Author: (It depends on whether we are trying to keep the "advanced" info somewhat hidden, to avoid scaring users with the most basic needs)

Reviewer 3: Ronnie suggested a "spolier" which hides the content.

Author: Yep I like that.

>       In this example, a failed task blocked the flow because downstream
>       tasks depended on its success. Triggering the failed task to run again
>       (if it succeeds this time) allows the original flow to continue.
> 
>       The flow will only continue downstream of a triggered task if it has not
>       already traversed that part of the graph.
> 
>       So if you trigger a past task that did not block the flow, only the triggered
>       task itself will run - unless you tell Cylc to start a new flow.

Reviewer 3: IMO this is too technical, developers will understand it, but users probably won't (flows are an advanced concept). We should be able to explain this more simply without having to mention flows.

E.G:

```suggestion
      In this example, a task has failed, as a result, some downstream tasks are not running.

      Triggering the failed task causes it to run again.
      If it succeeds, the downstream tasks will run and the workflow will continue.

      Note, any downstream tasks which have already run will not be re-run as a result.
```

Author: I thought you might say that :-) 

I don't think a "flow" is very technical. It's an intuitive concept with a very intuitive name. A single logical run of the workflow "flows" through the graph. 

Your edit does get pretty close to describing what's going on with using the word "flow" but it stops short of saying how you can rerun "tasks which have already run" if you actually need to do that - which sometimes you might need to do.

Author: The aforementioned NIWA user did need to - she was surprised that the flow did not continue after retriggering a succeeded task like it did on retriggering the failed one in the example.

Author: Also, it's fine (and good!) to have more detailed "advanced" info available where appropriate so long as it doesn't distract from the core basic info. That's why I tried to "hide" in a new tab.

What if we called it "advanced" rather than "explanation"?

Reviewer 3: > I don't think a "flow" is very technical. It's an intuitive concept with a very intuitive name. A single logical run of the workflow "flows" through the graph.

"Flows" are not part of the minimal terminology / concepts required to use Cylc and are not covered in the tutorial. You only learn about flows when trying to do `--flow=new`, most of our users do not know what "flow" means and there is great confusion amongst those who have tried to use use them.

But we don't need to talk about flows here, so lets avoid it.

Reviewer 3: ```
      In this example, a task has failed, as a result, some downstream tasks are not running.

      Triggering the failed task causes it to run again.
      If it succeeds, the downstream tasks will run and the workflow will continue.

      Note, any downstream tasks which have already run will not be re-run as a result.
      If you want to re-run downstream tasks, see :ref:`interventions.re-run-multiple-tasks`
```

You'll need to add a `interventions.re-run-multiple-tasks` anchor in the next section to use this suggestion.

Author: But you haven't addressed the core point that motivated this PR. I'll repeat it:

If we don't explain why the example works the way it does, users won't be able to apply their knowledge to slightly different use cases. In particular (and as motivated by an actual USER) they need to know (a) why rerunning a failed task typically has different "flow-on" effects than rerunning a succeeded task; and (b) how to make the succeed case flow on if they actually need that.

I don't think intervention documentation is the right place to stick to absolute basics. That's what the tutorial is for.

Author: (Our last two comments passed in the ether - I think you have mostly addressed my concerns now).

Author: > As there is only one flow in this example, we can explain this in the same way we would have at Cylc 7.

Yes and no - users also need to know why the workflow won't "flow on" after triggering a past task that did not block the flow.  

You have sort-of address this by referring forward to "re-run multiple tasks", but IMO it would help to have a brief explanation of the sort I suggested in this example too, because it helps put the specific example in a wider context. (After all the section is called "Re-run a Task" (any task!) not just "Re-run a Failed Task).
